### PR TITLE
Implement KMS methods for keys, policies and identities

### PR DESCRIPTION
**Pull request by reivaj05**
## Description
It adds several methods that will use KES client in minio to handle KES keys, policies and identities

## Motivation and Context
We need to expose KES in minio so we can use it later in Console
https://github.com/miniohq/engineering/issues/922

This is the first part of the integration, in following PR's, I'll expose these methods in new minio handlers so they can be accessed in minio

## How to test this PR?

Test can be done once handlers are added

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 15673-4aeb85a/gopath/src/github.com/minio
+ git -C 15673-4aeb85a/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 15673-4aeb85a/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 15673-4aeb85a/gopath/src/github.com/minio/minio fetch --quiet minio pull/15673/head:pr15673
+ git -C 15673-4aeb85a/gopath/src/github.com/minio/minio checkout --quiet pr15673
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/15673-4aeb85a/gopath
+ make -C 15673-4aeb85a/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/cli v1.23.0
go: downloading github.com/minio/kes v0.20.0
go: downloading golang.org/x/crypto v0.0.0-20220722155217-630584e8d5aa
go: downloading github.com/minio/minio-go/v7 v7.0.35
go: downloading golang.org/x/sys v0.0.0-20220722155257-8c9f86f7a55f
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/minio/pkg v1.3.2
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/Shopify/sarama v1.35.0
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/klauspost/compress v1.15.9
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/minio/madmin-go v1.4.26
go: downloading github.com/golang-jwt/jwt/v4 v4.4.1
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/coreos/go-oidc v2.1.0+incompatible
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/minio/console v0.20.1
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/go-openapi/loads v0.21.1
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/felixge/fgprof v0.9.2
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/miekg/dns v1.1.49
go: downloading github.com/klauspost/reedsolomon v1.10.0
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/minio/xxml v0.0.3
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/prometheus/client_golang v1.12.2
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/common v0.34.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/shirou/gopsutil/v3 v3.22.7
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.2
go: downloading go.etcd.io/etcd/api/v3 v3.5.4
go: downloading go.etcd.io/etcd/client/v3 v3.5.4
go: downloading go.uber.org/atomic v1.9.0
go: downloading golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5
go: downloading google.golang.org/api v0.81.0
go: downloading go.uber.org/zap v1.21.0
go: downloading github.com/coredns/coredns v1.9.3
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/go-ldap/ldap/v3 v3.4.3
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/ncw/directio v1.0.5
go: downloading golang.org/x/time v0.0.0-20220224211638-0e9765cccd65
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.1
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.8
go: downloading github.com/lib/pq v1.10.4
go: downloading github.com/nats-io/nats.go v1.16.0
go: downloading github.com/nats-io/stan.go v0.10.2
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.5.1
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/mattn/go-colorable v0.1.12
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/fraugster/parquet-go v0.10.0
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/klauspost/cpuid/v2 v2.1.0
go: downloading golang.org/x/net v0.0.0-20220722155237-a158d28d115b
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20211209120228-48547f28849e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/google/pprof v0.0.0-20211214055906-6f57359322fd
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/go-openapi/analysis v0.21.2
go: downloading github.com/go-openapi/spec v0.20.5
go: downloading github.com/go-openapi/swag v0.21.1
go: downloading github.com/lestrrat-go/jwx v1.2.19
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/fatih/structs v1.1.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading golang.org/x/sync v0.0.0-20220513210516-0976fa681c29
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.4
go: downloading google.golang.org/grpc v1.46.2
go: downloading cloud.google.com/go v0.100.2
go: downloading google.golang.org/genproto v0.0.0-20220519153652-3a47de7e79bd
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.0
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/rs/xid v1.4.0
go: downloading gopkg.in/ini.v1 v1.66.6
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/tklauser/go-sysconf v0.3.10
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/pierrec/lz4/v4 v4.1.15
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/tklauser/numcpus v0.4.0
go: downloading github.com/go-openapi/strfmt v0.21.2
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading go.mongodb.org/mongo-driver v1.9.1
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/mitchellh/mapstructure v1.4.3
go: downloading github.com/oklog/ulid v1.3.1
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/go-openapi/errors v0.20.2
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading cloud.google.com/go/iam v0.2.0
go: downloading github.com/googleapis/gax-go/v2 v2.4.0
go: downloading golang.org/x/xerrors v0.0.0-20220517211312-f3a8303e98df
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-stack/stack v1.8.1
go: downloading github.com/go-openapi/runtime v0.23.3
go: downloading github.com/go-openapi/validate v0.21.0
go: downloading github.com/unrolled/secure v1.10.0
go: downloading github.com/minio/mc v0.0.0-20220818165341-8c239d16aa37
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/tidwall/gjson v1.14.2
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/docker/go-units v0.4.0
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading github.com/minio/colorjson v1.0.2
go: downloading github.com/gdamore/tcell/v2 v2.4.1-0.20210905002822-f057f0a857a1
go: downloading github.com/charmbracelet/bubbletea v0.20.0
go: downloading github.com/charmbracelet/bubbles v0.10.3
go: downloading github.com/prometheus/prom2json v1.3.1
go: downloading github.com/posener/complete v1.2.3
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/rivo/tview v0.0.0-20220216162559-96063d6082f3
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/navidys/tvxwidgets v0.1.0
go: downloading github.com/pkg/xattr v0.4.5
go: downloading github.com/charmbracelet/lipgloss v0.5.0
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/muesli/termenv v0.11.1-0.20220212125758-44cd13922739
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading cloud.google.com/go/compute v1.6.1
# github.com/minio/minio/internal/kms
internal/kms/kes.go:173:70: undefined: KMSKey
internal/kms/kes.go:181:14: undefined: KMSKey
internal/kms/kes.go:184:27: undefined: KMSKey
internal/kms/single-key.go:106:55: undefined: KMSKey
make: *** [build] Error 2

```

**Comment by harshavardhana** [comment]
> go: downloading cloud.google.com/go/compute v1.6.1
> # github.com/minio/minio/internal/kms
> internal/kms/kes.go:173:70: undefined: KMSKey
> internal/kms/kes.go:181:14: undefined: KMSKey
> internal/kms/kes.go:184:27: undefined: KMSKey
> internal/kms/single-key.go:106:55: undefined: KMSKey
> make: *** [build] Error 2

@reivaj05 please push only changes that you know can be built otherwise we will have some cleanup to do on PRs being built into mint CI/CD

File: internal/kms/kms.go [hunk 1]
```diff
@@ -61,6 +79,41 @@ type KMS interface {
 	// by the key ID. The contexts must match the context value
 	// used to generate the ciphertexts.
 	DecryptAll(ctx context.Context, keyID string, ciphertext [][]byte, context []Context) ([][]byte, error)
+
+	// DescribePolicy describes a policy by returning its metadata.
```
Comment (klauspost) [inline internal/kms/kms.go]: For these, that are quite specific, would it make sense to add these as separate interfaces?

Bloating the `KMS` interface reduces the usability and forces stubs for other implementations .

Go Proverb: [The bigger the interface, the weaker the abstraction](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=5m17s)

Instead add `kms.KeyManager`, `kms.Policy` and `kms.Identity` interfaces could be added and the caller can assert them on the KMS for whatever operation they need.

Usage would look like this:
```
var globalKMS kms.KMS

func listPolicy(...) error {
    if k, ok := globalKMS.(kms.Policy); ok {
      k.DescribePolicy(...)
    }
    return errors.New("kms: describe policy is not supported")
}
```

... or alternatively:
```
var globalKMSPolicy kms.Policy

func listPolicy(...) error {
    if globalKMSPolicy == nil {
      return errors.New("kms: describe policy is not supported")
    }
    globalKMSPolicy.DescribePolicy(...)
}
```
... or you could have a dummy implementation of `kms.Policy` that just returns errors.

There are plenty of options where you don't need to pollute the interface.

File: internal/kms/kms.go [hunk 2]
```diff
@@ -89,6 +142,35 @@ type DEK struct {
 	Ciphertext []byte
 }
 
+// PolicyInfo describes a KMS policy.
+type PolicyInfo struct {
```
Comment (aead) [inline internal/kms/kms.go]: You can actually reuse these types: https://pkg.go.dev/github.com/minio/kes#PolicyInfo

Comment (aead) [reply internal/kms/kms.go]: So also `IdentityInfo` and others.

File: internal/kms/kms.go [hunk 3]
```diff
@@ -38,6 +39,16 @@ type KMS interface {
...
 
+	// DeleteKey deletes a key at the KMS with the given key ID.
+	// Please note that is a dangerous operation.
+	// Once a key has been deleted all data that has been encrypted with it cannot be decrypted
+	// anymore, and therefore, is lost.
+	DeleteKey(ctx context.Context, keyID string) error
+
+	// ListKeys List all key names that match the specified pattern. In particular,
+	// the pattern * lists all keys.
+	ListKeys(ctx context.Context, pattern string) ([]Key, error)
```
Comment (aead) [inline internal/kms/kms.go]: Ideally stick to the KES API which returns iterators for sets of unknown size:

Comment (aead) [reply internal/kms/kms.go]: https://pkg.go.dev/github.com/minio/kes#KeyIterator

Comment (aead) [reply internal/kms/kms.go]: Same for `ListPolicies` and `ListIdentities`.

File: internal/kms/kms.go [hunk 4]
```diff
@@ -38,6 +39,16 @@ type KMS interface {
 	// CreateKey creates a new key at the KMS with the given key ID.
 	CreateKey(ctx context.Context, keyID string) error
 
+	// DeleteKey deletes a key at the KMS with the given key ID.
```
Comment (aead) [inline internal/kms/kms.go]: Maybe we can split this into two KMS interfaces:
```
type KMS interface {
   // Here all methods implemented by KES and the static key KMS
}
```

```
type KES interface {
   // Here all KES-specific methods. 
} 
```

Later we can simply check whether `GlobalKMS` is a KES client via:
```
if kms, ok := GlobalKMS.(KES); ok {
   kms.ListKeys(ctx)
}
```

**Review by aead — DISMISSED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-erasure.sh | :x: [more...](#48b77912-a943-4787-adf2-a7d3a822cd95) |
<a name="48b77912-a943-4787-adf2-a7d3a822cd95"></a>
### 15673-32caa25/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s4:31590
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 7 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 52 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 10 seconds
{
  "name": "mc",
  "duration": "10184",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-1614 --quiet --no-color rb myminio/mc-test-bucket-11205  >>> \nmc: <ERROR> Unable to validate target `myminio/mc-test-bucket-11205`. Get \"http://minio-k8s4:31590/mc-test-bucket-11205/?location=\": dial tcp 15.15.15.4:31590: i/o timeout"
}
(7/14) Running minio-go tests ... done in 1 minutes and 21 seconds
(8/14) Running minio-java tests ... done in 21 seconds
(9/14) Running minio-js tests ... done in 59 seconds
(10/14) Running minio-py tests ... done in 52 seconds
(11/14) Running s3cmd tests ... done in 24 seconds
(12/14) Running s3select tests ... done in 2 seconds
(13/14) Running versioning tests ... done in 3 minutes and 1 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by reivaj05** [comment]
@klauspost @harshavardhana @aead Could you please help me to review this PR one more time?

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by klauspost — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#da3a9fe6-fed0-4d41-8002-c41ed2689847) |
<a name="da3a9fe6-fed0-4d41-8002-c41ed2689847"></a>
### 15673-85e4dd1/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s7:32001
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 58 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 27 seconds
(8/14) Running minio-go tests ... done in 1 minutes and 4 seconds
(9/14) Running minio-java tests ... done in 31 seconds
(10/14) Running minio-js tests ... done in 1 minutes and 3 seconds
(11/14) Running minio-py tests ... done in 1 minutes and 20 seconds
(12/14) Running s3cmd tests ... FAILED in 5 minutes and 19 seconds
{
  "name": "s3cmd",
  "duration": "306395",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "WARNING: Retrying failed request: / (timed out)\nWARNING: Waiting 3 sec...\nERROR: S3 error: 409 (BucketAlreadyOwnedByYou): Your previous request to create the named bucket succeeded and you already own it."
}
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
@klauspost this PR is rebased with master and seems like it's still failing on windows

```
--- PASS: TestGetObjectInlineNotInline (0.07s)
=== RUN   TestGetObjectWithOutdatedDisks
    erasure-object_test.go:1214: Test 1: Expected data to have md5sum = `23ca472302f49b3ea5592b146a312da0`, found `1fc602bb3c9540e6dc5c255a1c2e6592`
--- FAIL: TestGetObjectWithOutdatedDisks (0.12s)
```

**Review by aead — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :x: [more...](#c44750ce-5fd1-4120-831f-25d0e7d23b9b) |
<a name="c44750ce-5fd1-4120-831f-25d0e7d23b9b"></a>
### 15673-492d636/mint-pools.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s4:31795
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 55 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 30 seconds
(8/14) Running minio-go tests ... done in 1 minutes and 26 seconds
(9/14) Running minio-java tests ... done in 40 seconds
(10/14) Running minio-js tests ... FAILED in 1 minutes and 10 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 1,
  "status": "FAIL",
  "error": "Error: read ECONNRESET at TCP.onStreamRead (internal/stream_base_commons.js:209:20)"
}
(10/14) Running minio-py tests ... done in 1 minutes and 24 seconds
(11/14) Running s3cmd tests ... done in 25 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :x: [more...](#4f8c3df3-e4aa-4943-abda-3dd788c1e298) |
<a name="4f8c3df3-e4aa-4943-abda-3dd788c1e298"></a>
### 15673-28754ce/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s3:30952
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 52 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 29 seconds
(8/14) Running minio-go tests ... done in 1 minutes and 32 seconds
(9/14) Running minio-java tests ... FAILED in 2 minutes and 8 seconds
{
  "name": "minio-java",
  "function": "listObjects()",
  "args": "[bucket, prefix, recursive, 1050 objects]",
  "duration": 122041,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty, bucketName = minio-java-test-a1557s, objectName = null, resource = /minio-java-test-a1557s, requestId = 1716555C980B79D0, hostId = a2c67dd3-66c5-407f-a300-03ac96bf1c9d)\nrequest={method=DELETE, url=http://minio-k8s3:30952/minio-java-test-a1557s, headers=Host: minio-k8s3:30952\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.4.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20220919T180928Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20220919/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 326\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nStrict-Transport-Security: max-age=31536000; includeSubDomains\nVary: Origin\nVary: Accept-Encoding\nX-Amz-Request-Id: 1716555C980B79D0\nX-Content-Type-Options: nosniff\nX-Xss-Protection: 1; mode=block\nDate: Mon, 19 Sep 2022 18:09:28 GMT\n}\n >>> [io.minio.S3Base$1.onResponse(S3Base.java:690), okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519), java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), java.base/java.lang.Thread.run(Thread.java:829)]"
}
(9/14) Running minio-js tests ... done in 1 minutes and 1 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 19 seconds
(11/14) Running s3cmd tests ... done in 25 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Pull request merged**